### ISSUE TITLE
[question] Is amp CUDA-specific?

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#36169]
E.g. can GradScaler be  theoretically used for CPU-training with bfloat16 to detect underflow? (codes in https://github.com/pytorch/pytorch/tree/master/torch/cuda/amp seem to be not cuda-specific, and kernels in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/AmpKernels.cu seem quite generic as well)

If not, should it not be moved out of `torch.cuda` to a more global namespace (even if kernels for now exist for CUDA only)?

**Comment by gchanan** [comment]
CC @mcarilli.

**Comment by mcarilli** [comment]
Right now Amp is Cuda-specific.  The FP32 and FP16 oplists are based on the hardware capabilities of Volta and Turing GPUs and the properties of Pytorch's own Cuda kernels.

For different precisions/backends, I expect you could use the same Python-side API, but you'd probably need different sets of backend wrappers:  The "Smaller/Faster Type"  and FP32 oplists would need to be different, because the hardware capabilities and kernel implementations would be different.

bfloat16 has (almost) the same dynamic range as FP32, so it shouldn't require gradient scaling, but definitely needs the optimizer to act on FP32 weights to capture small updates.  An autocasting-style API where model leaves remain in default precision, and tensors are casted as they enter certain functions, is one way to ensure FP32 weight updates.

**Comment by vadimkantorov** [comment]
Thanks for your answer! It makes sense!

Maybe in future a better place for `GradScaler` is `torch.nn` or `torch.optim`...
